src/servers/http_server.cc

```diff
@@ -394,287 +433,318 @@ ReadDataFromJson(
-            &response_output_array, allocator, shape, shape_index + 1, base,
-            counter, dtype);
-        response_output_val->PushBack(response_output_array, allocator);
-      } else {
-        if (dtype != TRITONSERVER_TYPE_BYTES) {
-          rapidjson::Value data_val(base[*counter]);
-          response_output_val->PushBack(data_val, allocator);
-          *counter += 1;
-        } else {
-          uint32_t* len = reinterpret_cast<uint32_t*>(base + *counter);
-          char* cstr =
-              reinterpret_cast<char*>(base + *counter + sizeof(uint32_t));
-          rapidjson::Value data_val(cstr, *len, allocator);
-          response_output_val->PushBack(data_val, allocator);
-          *counter += *len + sizeof(uint32_t);
-        }
-      }
-    }
+  if (byte_size != expected_size) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL,
+        std::string(
+            "output tensor shape does not match size of output for '" +
+            output_name + "'")
+            .c_str());
   }
+
+  return nullptr;  // success
 }
 
 TRITONSERVER_Error*
 WriteDataToJson(
-    rapidjson::Value& response_output,
-    rapidjson::Document::AllocatorType& allocator, void* base)
+    TritonJson::Value* data_json, const std::string& output_name,
+    const TRITONSERVER_DataType datatype, const void* base,
+    const size_t byte_size, const size_t element_count)
 {
-  const rapidjson::Value& shape = response_output["shape"];
-  const char* dtype_str = response_output["datatype"].GetString();
-  const TRITONSERVER_DataType dtype = TRITONSERVER_StringToDataType(dtype_str);
+  switch (datatype) {
+    case TRITONSERVER_TYPE_BOOL: {
+      const uint8_t* bool_base = reinterpret_cast<const uint8_t*>(base);
+      if (byte_size != (element_count * sizeof(uint8_t))) {
+        return TRITONSERVER_ErrorNew(
+            TRITONSERVER_ERROR_INTERNAL,
+            std::string(
+                "output tensor shape does not match size of output for '" +
+                output_name + "'")
+                .c_str());
+      }
+      for (size_t e = 0; e < element_count; ++e) {
+        data_json->AppendBool((bool_base[e] == 0) ? false : true);
+      }
+      break;
+    }
 
-  rapidjson::Value data_array(rapidjson::kArrayType);
-  int counter = 0;
-  for (size_t i = 0; i < size_t(shape[0].GetInt()); i++) {
-    rapidjson::Value data_val(rapidjson::kArrayType);
-    switch (dtype) {
-      case TRITONSERVER_TYPE_BOOL: {
-        uint8_t* bool_base = reinterpret_cast<uint8_t*>(base);
-        WriteDataToJsonHelper(
-            &data_val, allocator, shape, 1, bool_base, &counter, dtype);
-        break;
+    case TRITONSERVER_TYPE_UINT8: {
+      RETURN_IF_ERR(WriteDataToJsonCheck(
+          output_name, byte_size, sizeof(uint8_t) * element_count));
+      const uint8_t* cbase = reinterpret_cast<const uint8_t*>(base);
+      for (size_t e = 0; e < element_count; ++e) {
+        data_json->AppendUInt(cbase[e]);
```

Comment:
Here and for all Appending of output data, don't we want to preserve the shape?

Reply:
So that the returned json array is *not* flattened like it is in this code? I'm ok returning flattened for now. We would need to add an option to allow users to select "non-flattened" if we wanted to support that.

Reply:
So for the 20.06 release do we intend on adding support for non-flat output tensors back (along with the option for users to ask for shaped outputs)?

Reply:
Not for 20.06. We needed non-flat input because that was part of the spec. But for outputs the spec doesn't really say or have a way to control how they are returned. If we get requests then we will add it.